google/gax/bundling.py

```diff
@@ -0,0 +1,252 @@
+        to be used in the id
+
+    Returns:
+      tuple: computed as described above
+
+    Raises:
+      AttributeError: if any descriminator fields attribute does not exist
+    """
+    return tuple(_str_dotted_getattr(obj, x) for x in descriminator_fields)
+
+
+class Task(object):
+    """Coordinates the execution of a single bundle."""
+
+    def __init__(self, api_call, bundle_id, bundled_field, bundling_request):
+        """Constructor.
+
+        Args:
+           api_call (callable[[object], object]): the func that is this tasks's
+             API call
+           bundle_id: the id of this bundle
+           bundled_field: the field used to create the bundled request
+           bundling_request: the request to pass as the arg to api_call
+
+        """
+        self._api_call = api_call
+        self._bundling_request = bundling_request
+        self.bundle_id = bundle_id
+        self.bundled_field = bundled_field
+        self.in_deque = collections.deque()
+        self.out_deque = collections.deque()
+
+    @property
+    def message_count(self):
+        """The number of bundled messages."""
+        return len(self.in_deque)
+
+    @property
+    def message_bytesize(self):
+        """The size of in bytes of the bundle messages."""
+        return sum(len(str(x)) for x in self.in_deque)
+
+    def run(self):
+        """Call the task's func.
+
+        The task's func will be called with the bundling requests func
+        """
+        if len(self.in_deque) == 0:
+            return
+        req = self._bundling_request
+        setattr(req, self.bundled_field, list(x for x in self.in_deque))
+        self.in_deque.clear()
+        try:
+            resp = self._api_call(req)
+            self.out_deque.append(resp)
+        except Exception as exc:  # pylint: disable=broad-except
+            self.out_deque.append(exc)
+
+    def append(self, msg):
+        """Adds a msg to the in_queue."""
+        self.in_deque.append(msg)
+
+    def canceller_for(self, msg):
+        """Obtains a cancellation function for msg.
+
+`        The returned cancellation function returns ``True`` if the message
+        was removed successfully from the in_deque, and false if it was not.
+
+
+        Args:
+           msg (object): the message to be cancelled
+
+        Returns:
+           (callable[[], boolean]): used to remove the message from the in_deque
+
+        """
+
+        def canceller():
+            """Cancels submission of ``msg`` as part of this bundle."""
+            try:
+                self.in_deque.remove(msg)
+                return True
+            except ValueError:
+                return False
+
+        return canceller
+
+
+TIMER_FACTORY = threading.Timer
+"""A class with an interface similar to threading.Timer.
+
+Defaults to threading.Timer.  This makes it easy to plug-in alternate
+timer implementations."""
+
+
+class Executor(object):
```

Comment:
I have Task class in this module and it's usual to have an Executor class that runs the Task, so I'd prefer to leave this as is.